src/python/pants/engine/internals/graph.py

```diff
@@ -198,34 +200,18 @@ class Owners:
 async def find_owners(owners_request: OwnersRequest) -> Owners:
     sources_set = FrozenOrderedSet(owners_request.sources)
     dirs_set = FrozenOrderedSet(os.path.dirname(source) for source in sources_set)
-
-    # Walk up the buildroot looking for targets that would conceivably claim changed sources.
     candidate_specs = tuple(AscendantAddresses(directory=d) for d in dirs_set)
-    candidate_targets = await Get[HydratedTargets](AddressSpecs(candidate_specs))
-
-    # Match the source globs against the expanded candidate targets.
-    def owns_any_source(legacy_target: HydratedTarget) -> bool:
-        """Given a `HydratedTarget` instance, check if it owns the given source file."""
-        target_kwargs = legacy_target.adaptor.kwargs()
-
-        # Handle `sources`-declaring targets.
-        # NB: Deleted files can only be matched against the 'filespec' (ie, `PathGlobs`) for a target,
-        # so we don't actually call `fileset.matches` here.
-        # TODO: This matching logic should be implemented using the rust `fs` crate for two reasons:
-        #  1) having two implementations isn't great
-        #  2) we're expanding sources via HydratedTarget, but it isn't necessary to do that to match
-        target_sources = target_kwargs.get("sources", None)
-        return target_sources and any_matches_filespec(
-            paths=sources_set, spec=target_sources.filespec
-        )
 
+    candidate_targets = await Get[Targets](AddressSpecs(candidate_specs))
     build_file_addresses = await MultiGet(
-        Get[BuildFileAddress](Address, ht.adaptor.address) for ht in candidate_targets
+        Get[BuildFileAddress](Address, tgt.address) for tgt in candidate_targets
     )
+
     owners = Addresses(
-        ht.adaptor.address
-        for ht, bfa in zip(candidate_targets, build_file_addresses)
-        if LegacyAddressMapper.any_is_declaring_file(bfa, sources_set) or owns_any_source(ht)
```

Comment:
It took me some time to realize what `LegacyAddressMapper` is doing. 

https://github.com/pantsbuild/pants/blob/9945df1b7dd02f866a0cfb7cc11ced6f82ae807d/src/python/pants/engine/legacy/address_mapper.py#L35-L41

Those first 4 lines are what matter. We are always passing it a `BuildFileAddress`, so we never execute the rest of the function. So, this PR just inlines that logic into the rule directly.